tests/Aspire.Hosting.Tests/MSBuildTests.cs

```diff
@@ -0,0 +1,138 @@
+        try
+        {
+            var libraryDirectory = Path.Combine(tempDirectory.FullName, "Library");
+            Directory.CreateDirectory(libraryDirectory);
+
+            File.WriteAllText(Path.Combine(libraryDirectory, "Library.csproj"), """
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+</Project>
+""");
+            File.WriteAllText(Path.Combine(libraryDirectory, "Class1.cs"), """
+namespace Library;
+
+public class Class1
+{
+}
+""");
+
+            var appHostDirectory = Path.Combine(tempDirectory.FullName, "AppHost");
+            Directory.CreateDirectory(appHostDirectory);
+
+            File.WriteAllText(Path.Combine(appHostDirectory, "AppHost.csproj"), $"""
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsAspireHost>true</IsAspireHost>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="{repoRoot}\src\Aspire.Hosting\Aspire.Hosting.csproj" AspireHostLibrary="true" />
+
+    <ProjectReference Include="..\Library\Library.csproj" />
+  </ItemGroup>
+
+</Project>
+""");
+            File.WriteAllText(Path.Combine(appHostDirectory, "Program.cs"), """
+var builder = DistributedApplication.CreateBuilder();
+builder.Build().Run();
+""");
+
+            File.WriteAllText(Path.Combine(appHostDirectory, "Directory.Build.props"), $"""
+<Project>
+  <PropertyGroup>
+    <SkipAspireWorkloadManifest>true</SkipAspireWorkloadManifest>
+    <_SuppressImportAspireHostingOrchestrationTargets>true</_SuppressImportAspireHostingOrchestrationTargets>
+  </PropertyGroup>
+
+  <Import Project="{repoRoot}\src\Aspire.Hosting\build\Aspire.Hosting.props" />
+</Project>
+""");
+            File.WriteAllText(Path.Combine(appHostDirectory, "Directory.Build.targets"), $"""
+<Project>
+  <Import Project="{repoRoot}\src\Aspire.Hosting\build\Aspire.Hosting.targets" />
+  <Import Project="{repoRoot}\src\Aspire.Hosting.Sdk\SDK\Sdk.targets" />
+</Project>
+""");
+
+            var output = new StringBuilder();
+            var outputDone = new ManualResetEvent(false);
+            var process = new Process();
+            // set '-nodereuse:false -p:UseSharedCompilation=false' so the MSBuild and Roslyn server processes don't hang around, which may hang the test in CI
+            process.StartInfo = new ProcessStartInfo("dotnet", $"build -nodereuse:false -p:UseSharedCompilation=false")
+            {
+                RedirectStandardOutput = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+                WorkingDirectory = appHostDirectory
+            };
+            process.OutputDataReceived += (sender, e) =>
```

Comment:
might it make sense to subscribe to ErrorDataReceived in case of some error? without introducing the [famous deadlock](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standarderror)

Reply:
basically just add `output.AppendLine(process.StandardError.ReadToEnd());`  just before WaitForExit.

Reply:
MSBuild doesn't write to stderr. See https://github.com/dotnet/msbuild/issues/3101.

I think we can add this in the future if it does start writing to stderr.